R/calc_TRI.R

```diff
@@ -0,0 +1,132 @@
+#' Calculate Terrain Ruggedness Index (TRI) based on SRTM rasters
+#'
+
+.calc_TRI <- function(shp,
+                      srtmelevation,
+                      engine = "zonal",
+                      stats = c('mean', 'median', 'sd'),
+                      rundir = tempdir(),
+                      verbose = TRUE,
+                      todisk = FALSE,
+                      ...) {
+
+  if (ncell(srtmelevation) > 1024 * 1024) todisk <- TRUE
+  if (engine == "extract") {
+    tib.zstats <- .comp_triExtract(
+      elevation = srtmelevation,
+      shp = shp,
+      stats = stats,
+      rundir = rundir,
+      todisk = todisk
+    )
+    return(tib.zstats)
+
+  } else {
+    tib.zstats <- .comp_triZonal(
+      elevation = srtmelevation,
+      shp = shp,
+      stats = stats,
+      todisk = todisk,
+      rundir = rundir
+    )
+    return(tib.zstats)
+  }
+}
+
+
+#' function to compute single raster with zonal
+#'
+
+.comp_triZonal <- function(elevation = NULL,
+                           shp = NULL,
+                           stats = c("mean", "median", "sd"),
+                           todisk = todisk,
+                           rundir = rundir,
+                           ...) {
+  shp.v <- vect(shp)
+  rast_mask <- terra::mask(elevation,
+    shp.v,
+    filename =  ifelse(todisk, file.path(rundir, "elevation.tif"), ""),
+    overwrite = TRUE
+  )
+  p_raster <- terra::rasterize(shp.v,
+    rast_mask,
+    field = 1:nrow(shp.v),
+    filename =  ifelse(todisk, file.path(rundir, "polygon.tif"), ""),
+    overwrite = TRUE
+  )
+  tri <- terra::terrain(rast_mask,
+    v = "TRI",
+    unit = "degrees",
+    neighbors = 8,
+    filename = ifelse(todisk, file.path(rundir, "terrain.tif"), ""),
+    overwrite = TRUE
+  )
+  tri.mean <- terra::zonal(tri,
```

Comment:
As noted in your comment it is best to include some if-statements to calculate mean/median/sd only if specified by the user. This also means that the stats specified should be checked for correctness at the top of the function

Reply:
Should we also give a try for `exact` engine using `exactextractr`?

Reply:
sure, another engine sounds great!